[feat] result types: add weather result answerer and template

Bnyro opened the pull request:
## What does this PR do?
- This PR adds result types and a template for weather engines

## Why is this change important?
- currently every engine hardcode its html

## How to test this PR locally?
- `!om Berlin`

## Author's checklist
- Duckduckgo Weather and Wttr.in are not migrated yet - they will be migrated as well after the answer type and template are finalized (if there any suggestions for improvements), I don't want to take the efforts to migrate and change them multiple times

## Related issues
superseeds/closes #3256
related: https://github.com/searxng/searxng/issues/4542, https://github.com/searxng/searxng/pull/1800, https://github.com/searxng/searxng/issues/1352

![2025-04-24T17:35:37,110507410+02:00](https://github.com/user-attachments/assets/b046a453-a418-409c-b3d1-1cd61359fd6a)
![2025-04-24T17:35:45,160382185+02:00](https://github.com/user-attachments/assets/afeff771-40ea-4f2a-84cd-37a3cdd6402b)

Bnyro committed:
[feat] result types: add weather result answerer and template

Bnyro committed:
[feat] open meteo: use new weather answerer

return42 committed:
[mod] implement searx.wikidata_units for unit converters

Bnyro commented on .pylintrc:
```diff
@@ -206,7 +206,7 @@ logging-modules=logging
 
 # List of note tags to take in consideration, separated by a comma.
-notes=FIXME,XXX,TODO
+notes=FIXME,XXX
```
The pylint changes are probably not related to this PR and should be separated?

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,610 @@
+    return WEATHER_DATA_CACHE
+
+
+def _get_sxng_locale_tag() -> str:
```
Maybe this should be moved to https://github.com/searxng/searxng/blob/master/searx/locales.py? It's not necessarily coupled to the specific weather engine request and could be useful in other places too.

return42 replied:
We have several places in the code where we try to determine the "ideal" language selection for the client. All of these implementations still differ in detail today. This mess will have to be cleaned up one day, but until then, it's more reliable to implement this method here rather than, for example, this one (which does something similar).

https://github.com/searxng/searxng/blob/848c8d0544eb0138da4414a750f79b22883f91b2/searx/webapp.py#L161-L164

I'm working on a parallel branch that cleans up the properties and also takes this problem into account to find a general solution.

Bnyro commented on searx/templates/simple/answer/weather.html:
```diff
@@ -0,0 +1,63 @@
+    {%- if answer.forecasts -%}
+      <div class="answer-weather-forecasts">
+        {%- for forecast in answer.forecasts -%}
+          <div class="summary">{{ forecast.datetime.l10n(locale=answer.current.location,format="short") }} {{ forecast.summary }}</div>
```
```suggestion
          <div class="summary">{{ forecast.datetime.l10n(locale=answer.current.location, fmt="short") }} {{ forecast.summary }}</div>
```
This currently leads to a crash, typo :)

Bnyro commented on searx/templates/simple/answer/weather.html:
```diff
@@ -0,0 +1,63 @@
+          {%- if data.url %}<img class="symbol" src="{{ data.url }}" title="{{ data.summary }}">{% endif -%}
+        </td>
+      </tr>
+      <tr>
+        <td>{{ _("Temperature") }}:</td>
+        <td>{{ data.temperature.l10n(locale=data.location) }}</td>
+        <td>{{ _("Feels Like") }}:</td>
+        <td>{{ data.feels_like.l10n(locale=data.location) }}</td>
+      </tr>
+      <tr>
+        <td>{{ _("Wind") }}:</td>
+        <td>{{ data.wind_from.l10n(locale=data.location) }}: {{ data.wind_speed.l10n(locale=data.location) }}</td>
+        <td>{{ _("Pressure") }}:</td>
+        <td>{{ data.pressure.l10n(locale=data.location) }}</td>
+      </tr>
+      <tr>
+        <td>{{_("Humidity")}}:</td>
+        <td>{{ data.humidity.l10n(locale=data.location) }}</td>
+        <td></td>
+        <td></td>
+      </tr>
```
Do we really want to make all these mandatory? I don't think all weather engines have pressure data (e.g. wttr.in I think).

return42 replied:
> Do we really want to make all these mandatory? 

I would leave the fields mandatory for now. If we switch to the other weather engines, we can always decide to relax this requirement.

However, every weather engine will have to meet a few minimum requirements, and I wouldn't want to decide which ones at this point (without having looked at all the other engines in depth).

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,610 @@
+        template: str = "{value} {unit}",
+        num_pattern: str = "#,##0",
+    ) -> str:
+        if isinstance(locale, GeoLocation):
+            locale = locale.locale()
+        elif locale is None:
+            locale = babel.Locale.parse(_get_sxng_locale_tag(), sep='-')
```
This could probably be moved to a helper function (or `_get_sxng_locale_tag()`)? It's the same in all `l10n` methods.

return42 replied:
In the currently implemented six classes, the default locale is determined from the query. However, we may eventually want to implement classes that determine their default locale solely from the UI property (where the language of the query is irrelevant). We can implement an additional helper function, as you suggested, but that would only be a wrapper that would then have to call [_get_sxng_locale_tag](https://github.com/searxng/searxng/pull/4663#discussion_r2106152038). That would be a few fewer lines of code, but I wasn't sure if it was worth it. After all, it would be another function.

If you find an additional helper function helpful, we can implement it.

Bnyro replied:
> If you find an additional helper function helpful, we can implement it.
>
You're right, it probably doesn't make too much sense looking at the code again. I've only seen that there's duplicated code and didn't really look deeper into why it's duplicated here :)

Bnyro commented on client/simple/src/less/weather.less:
```diff
@@ -0,0 +1,34 @@
+  div.summary {
+    margin: 0;
+    padding: 0.5rem 1rem;
+    background-color: var(--color-header-background);
```
Maybe we should add some rounded borders to the summary to make it fit in better?

```suggestion
    background-color: var(--color-header-background);
    .rounded-corners-tiny;
```
![2025-05-25T11:18:27,893244402+02:00](https://github.com/user-attachments/assets/6f201588-4823-4487-a4d0-8304ca3d83a3)

vs

![2025-05-25T11:26:27,171415336+02:00](https://github.com/user-attachments/assets/92c29225-7fdc-4a0c-b9f4-b9376c3c73a8)

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,610 @@
+        geo_props = cache.get(search_term, ctx=ctx)
+
+        if not geo_props:
+            geo_props = cls._query_open_meteo(search_term=search_term)
```
I'm not too sure if we really want to rely on OpenMeteo here, but I don't have a better idea either.

return42 replied:
This isn't set in stone either: if we have alternative sources, we can implement them alongside the `_query_open_meteo` method. Currently, however, I don't know of any alternatives, and open-meto responds quickly and robustly. Since we also cache the responses, this should provide us with a good initial solution.

Bnyro commented on searx/templates/simple/answer/weather.html:
```diff
@@ -0,0 +1,63 @@
+    <tbody>
+      <tr>
+        <td rowspan="4">
+          {%- if data.url %}<img class="symbol" src="{{ data.url }}" title="{{ data.summary }}">{% endif -%}
```
I think this should be in the bottom right of the result like for the other answerers and not in the summary line. Also, shouldn't this point you to the web ui of the weather engine with the location already entered instead of just to their docs? I think it's not really helpful for the user this way.

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,610 @@
+        return data_url
+
+    response = network.get(origin_url, timeout=3)
+    if response and response.status_code == 200:
```
```suggestion
    if response.status_code == 200:
```
I don't think `response` can ever be `None`?

Bnyro reviewed (commented):
Looks great already, thanks for your work!

I've left some comments / suggestions / questions.

return42 committed:
[mod] weather results: add types, i18n/l10n, symbols & unit conversions

The types necessary for weather information such as GeoLocation, DateTime,
Temperature,Pressure, WindSpeed, RelativeHumidity, Compass (wind direction) and
symbols for the weather have been implemented.

There are unit conversions and translations for weather property labels.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,607 @@
+    If symbol (SVG) is not already in the :py:obj:`WEATHER_DATA_CACHE` its
+    fetched from https://github.com/nrkno/yr-weather-symbols
+    """
+    # Symbols for darkmode/lightmode? .. and day/night symnbols? .. for latter
```
```suggestion
    # Symbols for darkmode/lightmode? .. and day/night symnbols? .. for later
```

return42 replied:
Its not the word "later" in sense of *time* ... its the word "latter" in sense of *the latest* in the list given before

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,607 @@
+    fetched from https://github.com/nrkno/yr-weather-symbols
+    """
+    # Symbols for darkmode/lightmode? .. and day/night symnbols? .. for latter
+    # we need a geopint (critical in sense of privacy)
```
What's a "geopint"?

return42 replied:
its a pint of beer waiting for you at a geopoint :-)

typo: geopint --> geopoint .. :facepalm: .. will fix it.

Bnyro commented on client/simple/src/less/weather.less:
```diff
@@ -0,0 +1,37 @@
+    margin: 0;
+    padding: 0.5rem 1rem;
+    background-color: var(--color-header-background);
+    .rounded-corners-tiny;
```
Currently the arrow to open the weather forecast is misaligned.

![2025-05-31T19:13:52,585033174+02:00](https://github.com/user-attachments/assets/d58c3413-13c8-451b-8e85-d14d48ce67d8)

Bnyro replied:
Okay, we can fix it by replacing the `<div class="summary">` with `<span class="summary">`, that's the simplest fix :+1:

return42 replied:
Your first suggestion "list-style: none" works better .. for now .. the `<summary>` element is the simplest solution... it's a bit impractical because you can only expand/collapse it in the header. We should perhaps turn it into a horizontal (left/right) slider in a future PR.

Bnyro commented on searx/weather.py:
```diff
@@ -0,0 +1,607 @@
+    If symbol (SVG) is not already in the :py:obj:`WEATHER_DATA_CACHE` its
+    fetched from https://github.com/nrkno/yr-weather-symbols
+    """
+    # Symbols for darkmode/lightmode? .. and day/night symnbols? .. for latter
```
```suggestion
    # Symbols for darkmode/lightmode? .. and day/night symbols? .. for the latter
```
:)

return42 committed:
[build] /static
